### DISCUSSION TITLE
Unusually large dynamicport range is configured for the GitHub Actions windows-latest machine

### DISCUSSION THREAD

**Discussion by ben-spiller** [community/community#26011] [Actions]
<p>Does anyone know why/if if it’s deliberate that the GitHub Actions “os: windows-latest” machine (non-docker) has an unusual configuration for the TCP dynamic (=ephemeral) port range?</p>
<p>As far as I know from the Microsoft docs all recent Windows versions are configured to follow the IANA standard port range for ephemeral ports, which gives plenty of ports for both ephemeral and server:</p>
<blockquote>
<p>netsh int ipv4 show dynamicport tcp<br>
Protocol tcp Dynamic Port Range<br>
Start Port      : 49152<br>
Number of Ports : 16384</p>
</blockquote>
<p>However on the GitHub Actions Windows machine it’s got a slightly mad (? :upside_down_face: ) configuration which assigns 100% of the ports to ephemeral use and leaves none for starting servers:</p>
<blockquote>
<p>netsh int ipv4 show dynamicport tcp<br>
Protocol tcp Dynamic Port Range<br>
Start Port      : 1024<br>
Number of Ports : 64511</p>
</blockquote>
<p>I couldn’t see anything in the doc about this, so was assuming Actions would be providing a fairly standard configuration. Is it intended or a bug? Would certainly be good to document if this machine has an unusual configuration.</p>
<p>I’m using GitHub Actions for a testing framework that needs to allocate lots of server ports to run tests of a server application, and our algorithm for allocating ports (excluding the ephemeral range) falls over with this configuration. I can use “netsh int ipv4 set dynamicportrange tcp” as a workaround but would like to understand if this atypical machine configuration was intended, and if so it’d be great to mention this (and any other unusual configuration choices) in the docuemntation.</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/ben-spiller">@ben-spiller</a>,</p>
<p>Thank you for reaching this out!</p>
<p>GitHub hosts Linux and Windows runners on Standard_DS2_v2 virtual machines in <code>Microsoft Azure</code> (doc <a href="https://docs.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners#cloud-hosts-for-github-hosted-runners" rel="nofollow noopener">here</a>). The configuration including port is consistent.</p>
<p>checked on Github:<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/5/57b7a511a982095558be744d7464a11ccd3a86b8.png" data-download-href="/uploads/short-url/cvYZOQRK4WfmJOJ0AOleGIhdPGo.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181094743-c306c2e9-d25b-47e7-b44e-bbc2287ecf5a.png" alt="image" data-base62-sha1="cvYZOQRK4WfmJOJ0AOleGIhdPGo" width="690" height="234" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/5/57b7a511a982095558be744d7464a11ccd3a86b8_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">751×255 9.21 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Checked with Azure ‘windows-2019’ machine:<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/4/47f29e9b67b27a4752710dc21df21cca8c315b83.png" data-download-href="/uploads/short-url/agtI2OpUlGKrxRRBbne6WfeM3mz.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181094745-6705e0f1-d091-4849-af92-3b2b0f575445.png" alt="image" data-base62-sha1="agtI2OpUlGKrxRRBbne6WfeM3mz" width="690" height="179" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/4/47f29e9b67b27a4752710dc21df21cca8c315b83_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">881×229 13.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Thanks</p>

**Comment by ben-spiller** [comment]
<p>Thanks for looking into this!</p>
<p>I just created a win2019 machine myself in azure (“Windows Server 2019 Datacenter”)<br>
and it’s using the standard IANA range:</p>
<h2>Protocol tcp Dynamic Port Range</h2>
<p>Start Port      : 49152<br>
Number of Ports : 16384</p>
<p>The MS docs for dynamicport are a bit sketchy but <a href="https://support.microsoft.com/en-gb/help/929851/the-default-dynamic-port-range-for-tcp-ip-has-changed-in-windows-vista" rel="nofollow noopener">https://support.microsoft.com/en-gb/help/929851/the-default-dynamic-port-range-for-tcp-ip-has-changed-in-windows-vista</a> seems to indicate the same IANA port range (dynamicport for 19152+) for modern windows versions (see the “applies to” which includes 2019).</p>
<p>I wonder why you’re seeing something different. How exactly did you create your win2019 Azure machine where you saw the 1024+ dynamic port range? Did you tick any non-default options? I did: “create a virtual machine”, selected my region (UK South), image=2019 gen1, size=DS1_v2 (DS2 option was not available), inbound port rules=allow selected (RDP i.e. the default). But I doubt any of those would make a difference to this.</p>
<p>I don’t suppose anyone reading this knows someone in the Azure or Windows Server teams who might be able to shed light on why some machines seem to have a different out-of-the-box dynamicport configuration?</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/ben-spiller">@ben-spiller</a>,</p>
<p>Thanks for your reply!</p>
<p>No special setting on my azure pipeline, just select windows-2019 and use a powershell script to get the info, this will use hosted vm for checking. (didn’t create VM)<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/8/86704437c2014c04fd53bbe38870f001a4d5cae4.png" data-download-href="/uploads/short-url/jbivFJ6UVK5V0duFsnhlcS1wUpC.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181094748-4d97ffcd-66c1-4cf8-a87c-b67f0b8863e3.png" alt="image" data-base62-sha1="jbivFJ6UVK5V0duFsnhlcS1wUpC" width="690" height="272" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/8/86704437c2014c04fd53bbe38870f001a4d5cae4_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">983×388 13.9 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Powershell script:</p>
<pre><code class="lang-auto">netsh int ipv4 show dynamicport tcp
echo ------------------
systeminfo | findstr /B /C:"OS Name" /C:"OS Version"
</code></pre>
<p>Could you please check on your side?</p>
<p>Thanks</p>

**Comment by al-cheb** [comment]
<p>Hi, <a class="mention" href="/ben-spiller">@ben-spiller</a><br>
We generate images for GitHub and Azure Pipelines using the <a href="https://github.com/actions/virtual-environments" rel="nofollow noopener">https://github.com/actions/virtual-environments</a> repository.</p>
<p>After installing VS via script <a href="https://github.com/actions/virtual-environments/blob/2602d388778a3626a5cba3f332bbf4283b4a5fa2/images/win/scripts/Installers/Install-VS.ps1" rel="nofollow noopener">https://github.com/actions/virtual-environments/blob/2602d388778a3626a5cba3f332bbf4283b4a5fa2/images/win/scripts/Installers/Install-VS.ps1</a> the dynamic port range is changed to:</p>
<pre><code class="lang-auto">Protocol tcp Dynamic Port Range
    vhd: ---------------------------------
    vhd: Start Port      : 1024
    vhd: Number of Ports : 64511
</code></pre>
<p>Example <a href="https://github.visualstudio.com/virtual-environments/_build/results?buildId=78654&amp;view=logs&amp;j=011e1ec8-6569-5e69-4f06-baf193d1351e&amp;t=486c0909-f67d-5748-80b3-bb1d531e7b43" rel="nofollow noopener">https://github.visualstudio.com/virtual-environments/_build/results?buildId=78654&amp;view=logs&amp;j=011e1ec8-6569-5e69-4f06-baf193d1351e&amp;t=486c0909-f67d-5748-80b3-bb1d531e7b43</a> :</p>
<p>Before Visual Studio installation [703]:</p>
<pre><code class="lang-auto">==&gt; vhd: Provisioning with powershell script: C:\agent\_work\9\s\images\win/scripts/Installers/Install-VS.ps1
    vhd: Display ports
    vhd:
    vhd: Protocol tcp Dynamic Port Range
    vhd: ---------------------------------
    vhd: Start Port      : 49152
    vhd: Number of Ports : 16384
</code></pre>
<p>After [825]:</p>
<pre><code class="lang-auto">    vhd: Display ports
    vhd:
    vhd: Protocol tcp Dynamic Port Range
    vhd: ---------------------------------
    vhd: Start Port      : 1024
    vhd: Number of Ports : 64511
    vhd:
</code></pre>
<p>We are currently discussing to revert ports to the default values or update the documentation (<a href="https://github.com/actions/virtual-environments/blob/2602d388778a3626a5cba3f332bbf4283b4a5fa2/images/win/Windows2019-Readme.md" rel="nofollow noopener">https://github.com/actions/virtual-environments/blob/2602d388778a3626a5cba3f332bbf4283b4a5fa2/images/win/Windows2019-Readme.md</a>) about that moment.</p>

**Comment by ben-spiller** [comment]
<p>Great thanks for doing that analysis!</p>
<p>Wow this is quite surprising that installating Visual Studio 2019 seems to have changed the machine’s dynamicport TCP configuration. I couldn’t find any VS documentation about that, it smells like possibly an unintentional change/bug?</p>
<p>Either way I guess the <a href="https://github.com/actions/virtual-environments/blob/9110b389998666dc0c6a8e9abd9400248bad8dba/images/win/scripts/Installers/Install-VS.ps1" rel="nofollow noopener">https://github.com/actions/virtual-environments/blob/9110b389998666dc0c6a8e9abd9400248bad8dba/images/win/scripts/Installers/Install-VS.ps1</a> could perhaps revert back to the default Windows configuration to avoid surprising people with this more limited TCP configuration that’s not good for server processes.</p>
<p>Anyway thanks for looking into it, I look forward to hearing what you come up with :slight_smile:</p>

**Comment by al-cheb** [comment]
<p><a class="mention" href="/ben-spiller">@ben-spiller</a>, I have prepared the <a href="https://github.com/actions/virtual-environments/pull/1442" rel="nofollow noopener">https://github.com/actions/virtual-environments/pull/1442</a> PR to revert the default value.</p>

**Comment by al-cheb** [comment, marked as answer]
<p><a class="mention" href="/ben-spiller">@ben-spiller</a>, The image with dynamic port changes has been deployed.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/c/cc488b8d2ee882a8f7844ee2e114dd9c5264a19d.png" data-download-href="/uploads/short-url/t9aPx3G41dTFVua1fWRbAUONXn7.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/47745270/181094752-a2899714-3012-4572-91d0-e405f30b1832.png" alt="image" data-base62-sha1="t9aPx3G41dTFVua1fWRbAUONXn7" width="664" height="500" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/c/cc488b8d2ee882a8f7844ee2e114dd9c5264a19d_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">893×672 23 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by ben-spiller** [comment]
<p>Awesome thanks, I’ve removed my workaround and now it’s working as expected. Thanks for the quick turnaround :slight_smile:</p>

**Comment by ben-spiller** [comment]
<p>Seems quite dodgy that Visual Studio is changing the machine’s TCP config in this radical way (without any documentation I’m aware of to let user’s know) but I guess that’s a problem for someone else to solve</p>
